Estimation of selective pressure on protein-coding sequences is a key comparative genomics approach for de novo prediction of lineage specific adaptations. Selective pressure is measured on a per gene basis by comparing the rate of non-synonymous substitutions to the rate of neutral evolution, typically assumed to be the rate of synonymous substitutions. We previously demonstrated that under time-reversible models of nucleotide evolution, the number of substitutions is systematically biased toward overestimation when sequence composition changes.

We extend these findings to the case of time-reversible codon substitution models and demonstrate overestimation of selective pressure (ie exaggerating the magnitude of purifying natural selection) over whole exome-scale data drawn from mammals, insects, frog, and fish. All published codon substitution models have been time-reversible and thus assume that sequence composition does not change over time. Our basis for comparison is a non-stationary codon substitution model that allows sequence composition to change.

We demonstrate by application of model selection techniques that our new codon model tends to fit the data better. We further show through direct measurement of non-stationarity that bias in selective pressure and rate of neutral evolution increase with the extent to which non-stationarity is violated.

With an increased rate of genomic sequencing across the tree of life, the de novo estimation of natural selection assumes even greater importance for identifying functionally important genes and processes. We have shown that the inference drawn under time-reversible models is affected by compositional divergence. Our new codon model provides a foundation for more robust inferences.

Estimation of selective pressure on protein-coding sequences is a key comparative genomics approach for de novo prediction of lineage specific adaptations. Selective pressure is measured on a per gene basis by comparing the rate of non-synonymous substitutions to the rate of neutral evolution, typically assumed to be the rate of synonymous substitutions. We previously demonstrated that under time-reversible models of nucleotide evolution, the number of substitutions is systematically biased toward overestimation when sequence composition changes.

We extend these findings to the case of time-reversible codon substitution models and demonstrate overestimation of selective pressure (ie exaggerating the magnitude of purifying natural selection) over whole exome-scale data drawn from mammals, insects, frog, and fish. All published codon substitution models have been time-reversible and thus assume that sequence composition does not change over time. Our basis for comparison is a non-stationary codon substitution model that allows sequence composition to change.

We demonstrate by application of model selection techniques that our new codon model tends to fit the data better. We further show through direct measurement of non-stationarity that bias in selective pressure and rate of neutral evolution increase with the extent to which non-stationarity is violated.

With an increased rate of genomic sequencing across the tree of life, the de novo estimation of natural selection assumes even greater importance for identifying functionally important genes and processes. We have shown that the inference drawn under time-reversible models is affected by compositional divergence. Our new codon model provides a foundation for more robust inferences.
