\documentclass[a4paper,oneside,12pt]{article}
\include{definitions}
\begin{document}
\title{$\omega$ is $K_a/K_s$}
\author{Ben Kaehler}
%\date{26 March 2012}
\maketitle
We show that the definition of $K_a/K_s$ used in \citet*{goldman1994codon} transfers unscathed to the context of non-stationary Markov models. Further, it is equivalent to the parameter $\omega$ in the following standard definition \citep*[see, for example,][p.~59]{yang2006computational}

Define the instantaneous transition rates of a codon substitution process as
\begin{align*}
q_{ij} = \begin{cases} 0, &\text{codons differ at more than one position},\\
r_{ij}, &\text{one synonymous difference between codons},\\
\omega r_{ij}, &\text{one non-synonymous difference between codons},\end{cases}
\end{align*}
where $r_{ij}$ is some representation of an evolutionarily neutral process and $\omega$ summarises the influence of selective pressure. Note that $r_{ij}$ is allowed to vary by transition in this formulation but has traditionally been further constrained to enforce, for instance, time-reversibility of the process. We denote the marginal codon probability row vector $\pi(t)$ for time $t>0$.

Define two new transition matrices:
\begin{align*}
q_{ij}^\text{S} &= \begin{cases}
r_{ij}, &\text{one synonymous difference between codons},\\
0, &\text{otherwise},\end{cases} \\
\text{and} \\
q_{ij}^\text{N} &= \begin{cases}
r_{ij}, &\text{one non-synonymous difference between codons},\\
0, &\text{otherwise},\end{cases}
\end{align*}
so that $Q = \omega Q^\text{N} + Q^\text{S}$, where $Q$, $Q^\text{N}$, and $Q^\text{S}$ are the matrices comprised of $q_{ij}$, $q_{ij}^\text{N}$, and $q_{ij}^\text{S}$, respectively, for their off-diagonal elements. Diagonal elements are calculated in the usual way for Markov generators.

Following \citet{goldman1994codon}, we define $\rho_s$ and $\rho_a$ to be the rates of synonymous and non-synonymous substitution per codon. They can be shown to be
\begin{align*}
		\rho_s = -\pi(t)\diag Q^\text{S} \quad\text{and}\quad \rho_a = -\omega\pi(t)\diag Q^\text{N}.
\end{align*}
Further, we define $\rho_s^1$ and $\rho_a^1$ as the potentials of of synonymous and non-synonymous substitutions when no selective constraints exist at the amino acid level, that is $\rho_s^1 = -\pi(t)\diag Q^\text{S}$ and $\rho_a^1 = -\pi(t)\diag Q^\text{N}$. These parameters correspond to $\rho_s^\infty$ and $\rho_a^\infty$ in \citet{goldman1994codon}. Likewise, $3\rho_s^1$ and $3\rho_a^1$ represent the numbers of synonymous and non-synonymous nucleotide sites per codon. So, the numbers of synonymous substitutions per synonymous site and non-synonymous substitutions per non-synonymous site are $K_s = \rho_s/3\rho_s^1 = 1/3$ and $K_a=\rho_a/3\rho_a^1 = \omega/3$. It is therefore clear that
\begin{align*}
		\omega = \frac{K_a}{K_s},
\end{align*}
which shows that the usual definition and understanding of $\omega$ is sufficient in the context of non-stationary Markov models.

\bibliographystyle{chicagoa}
\bibliography{papers}
\end{document}
